extensions/amp-base-carousel/1.0/scroller.js

```diff
@@ -0,0 +1,233 @@
+ * scrolling, things are immediately centered again. Since there can be some
+ * delay between scroll events, and we do not want to interrupt a scroll with a
+ * render, it cannot be too small. 200ms seems to be around the lower limit for
+ * this value on Android / iOS.
+ */
+const RESET_SCROLL_REFERENCE_POINT_WAIT_MS = 200;
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function Scroller(props) {
+  const {
+    children = [],
+    loop,
+    ignoreProgrammaticScroll,
+    restingIndex,
+    setRestingIndex,
+  } = props;
+
+  /**
+   * The number of slides we want to place before the
+   * reference or resting index. Only needed if loop=true.
+   */
+  const pivotIndex = Math.floor(children.length / 2);
+
+  /**
+   * The dynamic position that the slide at the resting index
+   * is with respect to its scrolling order. Only needed if loop=true.
+   */
+  const offsetRef = useRef(restingIndex);
+  const containerRef = useRef();
+  const slides = Slides({
+    children,
+    restingIndex,
+    container: containerRef,
+    offsetRef,
+    pivotIndex,
+    loop,
+  });
+  const currentIndex = useRef(restingIndex);
+
+  // Note: this has to be useLayoutEffect, not useEffect.
+  useLayoutEffect(() => {
+    if (!containerRef.current) {
+      return;
+    }
+    const container = containerRef.current;
+    ignoreProgrammaticScroll.current = true;
+    container./* OK */ scrollLeft = loop
+      ? container./* OK */ offsetWidth * pivotIndex
+      : container./* OK */ offsetWidth * restingIndex;
+  }, [ignoreProgrammaticScroll, restingIndex, pivotIndex, loop]);
+
+  // Trigger render by setting the resting index to the current scroll state.
+  const resetScrollReferencePoint = () => {
+    // Check if the resting index we are centered around is the same as where
+    // we stopped scrolling. If so, we do not need to move anything.
+    if (currentIndex.current === restingIndex) {
+      return;
+    }
+    setRestingIndex(currentIndex.current);
+  };
+
+  const debouncedResetScrollReferencePoint = debounce(
+    window,
+    resetScrollReferencePoint,
+    RESET_SCROLL_REFERENCE_POINT_WAIT_MS
+  );
+
+  // Track current slide without forcing render.
+  // This is necessary for smooth scrolling because
+  // intermediary renders will interupt scroll and cause jank.
+  const updateCurrentIndex = () => {
+    const container = containerRef.current;
+    const slideOffset = Math.round(
+      (container./* OK */ scrollLeft -
+        offsetRef.current * container./* OK */ offsetWidth) /
+        container./* OK */ offsetWidth
+    );
+    currentIndex.current = mod(slideOffset, children.length);
+  };
+
+  const handleScroll = () => {
+    if (ignoreProgrammaticScroll.current) {
+      ignoreProgrammaticScroll.current = false;
+      return;
+    }
+    updateCurrentIndex();
+    debouncedResetScrollReferencePoint();
+  };
+
+  return (
+    <>
+      <style>{styles.hideScrollbarPseudo}</style>
```

Comment:
This is not ideal but the only way I could think to include styling via pseudo-selector for now. Any thoughts on how to do it either or both of: (1) including the style with an alternative, cleaner approach, (2) isolating the style to once at the doc-level rather than with the carousel at the instance-level?

Reply:
Discussed offline, but (2) won't work at doc-level in AMP mode since the component is inside the shadow.